### DISCUSSION TITLE
How can I force a specific version of the Razor .cshtml editor in VS?

### DISCUSSION THREAD

**Discussion by daiplusplus** [dotnet/roslyn#54068] [Q&A] [answered]
I first asked this on gitter, but @CyrusNajmabadi recommended I post this here

> the reason i recommend Roslyn though is that Roslyn includes the razor editor team  
> so i presumed we (i'm on that team) would know the best about how this works :)

----------------------

Backstory: I'm the maintainer of RazorGenerator, the _embarrassingly slow moving_ NuGet package and VSIX that enables Razor use in non-ASP.NET projects, and I use RazorGenerator almost daily.

My problem is that when I'm using RazorGenerator with a .NET Core (non-ASP.NET Core) project then any `*.cshtml` files inside the project are opened by VS's Razor editor in "ASP.NET Core-mode" ("flavor"?) where it no-longer recognizes older Razor features like `@helper` and the Error List fills with squggies, even though the project build process completes without errors.

I understand that VS 2019 loads either the older V3 Razor editor (`Microsoft.WebTools.Languages.Razor.V3`, I think?) - or the newer Razor-for-ASP.NET Core editor (`Microsoft.WebTools.Languages.Razor.Core`?) based on the project type (.NET Framework vs. .NET Core), though I'm unsure how it chooses which flavor/mode to use when it's a .NET Standard project, or an ASP.NET Core 2.1 project running on .NET Framework 4.7+ - or other situations.

But I want to know how to force VS to open a `.cshtml` file in the Razor V3 editor instead of the Razor Core editor - or vice-versa depending on the situation.

-----

After some digging inside VS's assemblies, I've found `Microsoft.WebTools.Languages.Html.Editor.ContainedLanguage.Razor.Def.IRazorVersionDetector` and only one implementation: `Microsoft.WebTools.Languages.Html.VS.Razor.RazorVersionDetector`. I see that it checks for the MSBuild property `UseRazorV4Editor` but when I set `<UseRazorV4Editor>false</UseRazorV4Editor>` in my csproj it didn't have any effect.

**Comment by CyrusNajmabadi** [comment]
@NTaylorMullen Do you know the answer to the above?

**Comment by daiplusplus** [comment, marked as answer]
After spending too much time in Ghidra, I found the magic incantation...

So `RazorVersionDetector` (in `Microsoft.WebTools.Languages.Html.VS.dll`, in namespace `Microsoft.WebTools.Languages.Html.VS.Razor`),  performs these checks in sequence:

1. Does the filename end with `.vbhtml`? If so, then don't use the Razor Core editor.
1. Does the project have the `SupportsSystemWeb` capability? Or does the project *not* have the `.NET&CPS` capability?
   1.  If either of the above is true then check for the `UseRazorV4Editor` MSBuild property.
     1.  If `UseRazorV4Editor` is undefined or not set to `true` then use the ASP.NET Web Pages version (described below)
     1.  If `UseRazorV4Editor` is `true` then always use the Razor Core editor.
1. Then check the latest installed ASP.NET Web Pages configuration - if available.
    * On my computer at least, ASP.NET Web Pages is version `3.0.526.0`.
1. Then use the major version of ASP.NET Web Pages as the basis for the Razor version:
  * ASP.NET Web Pages v1, v2, and v3 will cause Razor V3 to be used, so `3.0.526.0` means Razor V3 will be used.
  * If ASP.NET Web Pages 4.0 is ever a thing then that will also cause Razor Core (aka Razor V4) to be used.

--------

So then it's just a matter of editing my .NET Core class library's `.csproj` to have this:

```
<Project Sdk="Microsoft.NET.Sdk">
	<PropertyGroup>
		<TargetFramework>netcoreapp3.1</TargetFramework>
		<SignAssembly>true</SignAssembly>
		<Nullable>enable</Nullable>
		<UseRazorV4Editor>false</UseRazorV4Editor>
	</PropertyGroup>
	<ItemGroup>
		<ProjectCapability Include="SupportsSystemWeb" />
	</ItemGroup>

	<!--  etc-->
</Project>
```

Now whenever I open ` .cshtml` file from within the project via the Solution Explorer it opens in the Razor V3 editor where it correctly supports `@helper`, phew!

**Reply by NTaylorMullen** [reply]
Bravo, you nailed the flow! Couldn't have said it better myself.

**Reply by daiplusplus** [reply]
@NTaylorMullen This is undocumented behaviour though. What can I do to ensure this will still work when a future Razor V5 editor comes out?

Also, if I understand it correctly it also means that a single project cannot target ASP.NET Core while also using RazorGenerator (and the V3 Razor editor ) for _some_ `.cshtml` files in the same project. As a workaround for that I think I could register a new file-extension (say, `.cshtml3` or `.v3cshtml`? are multi-dotted extensions like `.v3.cshtml` supported by the VS editor picker?) which then always uses the older extension? (Ideally I'd like to be able to detect RazorGenerator's asterisk-directives at the start of the `.cshtml` file and use that to force a specific editor to be used - but I don't think that's possible at all given how `RazorVersionDetector` is not designed to support any extensibility - or can a VSIX override the MEF `[Export( typeof(IRazorVersionDetector) )]` in this case?

Also, do you know why the V3 Razor editor in VS sometimes generates errors in the error-list such that double-clicking them does nothing (despite their associated filename and line-number being valid)?

**Reply by NTaylorMullen** [reply]
> @NTaylorMullen This is undocumented behaviour though. What can I do to ensure this will still work when a future Razor V5 editor comes out?

Ya it's definitely undocumented as of today so sadly I can't give you an absolute on how to build your stack in a way that will continue to work for future Razor. We're still landing on what that story will look like.

> As a workaround for that I think I could register a new file-extension (say, .cshtml3 or .v3cshtml?

That sounds like a fine workaround.

> are multi-dotted extensions like .v3.cshtml supported by the VS editor picker?) which then always uses the older extension?

I don't believe so 

> or can a VSIX override the MEF `[Export( typeof(IRazorVersionDetector) )]` in this case?

Tecccchnnnically you can do the same thing we do to take over the Razor editor in the form of a VS extension: https://github.com/dotnet/aspnetcore-tooling/blob/main/src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/RazorFilePathToContentTypeProvider.cs

It's definitely a hammer but it's a possible route.

**Comment by KristianSchmidt** [comment]
Hi @Jehoel and @NTaylorMullen

Apologies for the thread necromancy here, but we're seeing this issue pop up again with the release of VS2022.

We have a solution that runs on the .NET Framework 4.7.2, ASP.NET MVC 4 with the SDK-style project files. We did the upgrade to the SDK-style csprojs earlier this year, and by adding `<ProjectCapability Include="SupportsSystemWeb" />` we got Razor Intellisense working again, which this thread tells me comes from the choice of Razor editor within VS.

Now, with VS2022, it seems like it always chooses the Razor V4 editor. The error list is filled with `CS0234 The type or namespace name 'AspNetCore' does not exist in the namespace 'Microsoft' (are you missing an assembly reference?)`. Opening the same solution in VS2019, without changing anything, still works fine.

Ok, so in the spirit of doing your homework before asking questions, I tried a couple of different things.

1. I decompiled `Common7\IDE\Extensions\Microsoft\Web Tools\Languages\Microsoft.WebTools.Languages.Html.VS.dll`, finding the `RazorVersionDetector`, from both VS2019 and VS2022 respectively using dotPeek, and the logic inside these are the same as described by @Jehoel above.

2. I made a console app that references `Microsoft.VisualStudio.Web.dll` and prints out `AspNetWebPageConfiguration.LatestWebPagesVersion`, which is the value checked for the version of ASP.NET Web Pages. On my machine this did not return anything. Well, actually I got it to return 1.0 after installing some very old tooling I found on the MS website, but this did not seem to matter in VS2022.
But... I'm unsure if this logic actually needs to be run inside of VS to get the proper result. There might be some missing context when evaluating from an outside console app.

3. Digging deeper, I fired up WinDbg (Preview), deleted the VS component cache, ran devenv.exe, and tried to set a breakpoint inside any of the methods in `Microsoft.WebTools.Languages.Html.VS.Razor.RazorVersionDetector`. I was ultimately unsuccessful, and it was a quite confusing experience, because many VS dlls seem to be NGEN'd, having a version suffixed with `.ni.dll` that also gets loaded. So it was a bit difficult to tell if I needed to set a managed breakpoint or an unmanaged one.
I tried both, but nothing got hit. Which was also very confusing to me, because I could see the function `RazorVersionDetector.GetDotNetCoreRazorVersion` being loaded but _not_ JITed, and then later having been JITed (i.e. run) without my breakpoint being hit. Oh well, could also be that my windbg-fu is a bit lacking.

4. Played around with the template for creating a VS Extension, seeing if it would be possible to create something that calls into `RazorVersionDetector`, but my lack of experience with the VS extension object model caused me to give up fairly quickly.

So, any information or help would be very much appreciated, and I'm up for trying any ideas you might have. :)

**Reply by CZEMacLeod** [reply]
I seem to remember a 'throw away' reference to the fact that Razor now uses the LSP only in VS2022 as of its release (in the launch event).
It used to be behind a preview flag in VS2019, but that is now gone; maybe the new mechanism doesn't respect this logic at all?

**Reply by daiplusplus** [reply]
That's concerning, if true - because Razor V3 and V4 are different, and V3 supports features that V4 doesn't (like `@helper` syntax, and configuration in `web.config` instead of just `RazorStart`).

I don't have VS2022 myself right now, but I'll investigate shortly.

**Reply by KristianSchmidt** [reply]
For what it's worth I can get the Razor V3 syntax to work in VS2022 just by using an old-style project file. There's also still support for creating these projects in VS2022.

**Reply by KristianSchmidt** [reply]
Aha, looks like I found a workaround! Go to Options -> Text Editor -> HTML -> Advanced and select "Use legacy Razor editor for ASP.NET Core". Restart VS and then it works 🎉 

I guess the downside is that this is probably a VS-wide setting, so if you jump between projects you might have to change this often.

Decompiling `Microsoft.WebTools.Languages.Html.Editor.dll` in both VS2019 and VS2022 reveals that this is indeed a new setting in VS2022. Only in the new version is `HtmlSettings.UseLegacyASPNETCoreRazorEditor` present.

**Reply by KristianSchmidt** [reply]
Small update: I managed to get dnSpy to break inside `RazorVersionDetector` only _after_ enabling `UseLegacyASPNETCoreRazorEditor`. So maybe this was needed all along to hit this code path.

**Reply by NTaylorMullen** [reply]
@KristianSchmidt if you create your own `IFilePathToContentTypeProvider` you should be able to set the content type to the legacy Razor content type to enable your scenario. For instance [here's our CSHTML detector](https://github.com/dotnet/razor-tooling/blob/bffca4810bec00b4a6421f3776acf509de8d8bda/src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/CSHTMLFilePathToContentTypeProvider.cs) that moves the content type to the new Razor LSP content type.

**Reply by CZEMacLeod** [reply]
@NTaylorMullen Could the built in cshtml (and presumably vbhtml) detectors not do the checks that previously worked. Perhaps excluding projects that have a project capability of `SupportsSystemWeb`? 

I'm not sure I want to turn off the feature in options, or have to install a custom extension to get what used to work, working again.
This means I cannot easily work with new projects in razor v4 with the new LSP engine, and old projects with razor v3 in class libraries using the correct V3 engine. 

This scenario is applicable to both [CZEMacLeod/MSBuild.SDK.SystemWeb](https://github.com/CZEMacLeod/MSBuild.SDK.SystemWeb) and [RazorGenerator/RazorGenerator](https://github.com/RazorGenerator/RazorGenerator) - in the latter case you might not have the extension installed if you use the [RazorGenerator.MsBuild](https://www.nuget.org/packages/RazorGenerator.MsBuild) nuget package, and the former case is a SDK only and has no extension. (I don't believe you can put a `IFilePathToContentTypeProvider` class in a NuGet package to only affect specific projects it is added to.)

**Reply by NTaylorMullen** [reply]
> @NTaylorMullen Could the built in cshtml (and presumably vbhtml) detectors not do the checks that previously worked. Perhaps excluding projects that have a project capability of `SupportsSystemWeb`?
> 
> I'm not sure I want to turn off the feature in options, or have to install a custom extension to get what used to work, working again. This means I cannot easily work with new projects in razor v4 with the new LSP engine, and old projects with razor v3 in class libraries using the correct V3 engine.

Maybe I missed a detail. The new editor shouldn't be turning itself on for v4 or v3 Razor projects. 

> I don't believe you can put a `IFilePathToContentTypeProvider` class in a NuGet package to only affect specific projects it is added to

You cannot

**Reply by CZEMacLeod** [reply]
@NTaylorMullen I think the problem is that the old workarounds listed by @Jehoel in the post https://github.com/dotnet/roslyn/discussions/54068#discussioncomment-865478 which uses `RazorVersionDetector` no longer work in the VS2022 release, and cshtml files (in SDK style projects?) always go to the LSP service which only supports Razor V4, not Razor V3 (or below) unless you manually switch Options -> Text Editor -> HTML -> Advanced and select "Use legacy Razor editor for ASP.NET Core".

The ideal scenario is that you can use the new LSP for ASP.NET Core projects, and the old editor in other projects, preferably by setting a project property and/or capability.

**Reply by NTaylorMullen** [reply]
> @NTaylorMullen I think the problem is that the old workarounds listed by @Jehoel in the post [#54068 (comment)](https://github.com/dotnet/roslyn/discussions/54068#discussioncomment-865478) which uses `RazorVersionDetector` no longer work in the VS2022 release, and cshtml files (in SDK style projects?) always go to the LSP service which only supports Razor V4, not Razor V3 (or below) unless you manually switch Options -> Text Editor -> HTML -> Advanced and select "Use legacy Razor editor for ASP.NET Core".
> 
> The ideal scenario is that you can use the new LSP for ASP.NET Core projects, and the old editor in other projects, preferably by setting a project property and/or capability.

Ahhh I see, if you all would be willing to contribute an extra capability check to [our LSP feature detector](https://github.com/dotnet/razor-tooling/blob/bffca4810bec00b4a6421f3776acf509de8d8bda/src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/DefaultLSPEditorFeatureDetector.cs#L113) we'd definitely be willing to look it over!

**Reply by CZEMacLeod** [reply]
@NTaylorMullen Had a look at the code
The issue looks like your definition of 
```c#
private const string DotNetCoreCSharpCapability = "CSharp&CPS";
```
which only detects that the project is using a SDK style project, not that it is dotnetcore (since you can target net4x).
Looking at https://github.com/dotnet/sdk/blob/main/src/WebSdk/ProjectSystem/Targets/Microsoft.NET.Sdk.Web.ProjectSystem.targets there is a capability defined of `AspNetCore`
Alternatively https://github.com/dotnet/sdk/blob/main/src/RazorSdk/Targets/Microsoft.NET.Sdk.Razor.DesignTime.targets suggests `DotNetCoreRazor`

Unfortunately, I don't have time to directly check these just now but I think changing to 
```c#
private const string DotNetCoreCSharpCapability = "CSharp&AspNetCore";
```
or
```c#
private const string DotNetCoreCSharpCapability = "CSharp&DotNetCoreRazor";
```
would resolve the problem.

I am aware of the capability `DotNetCoreWeb` but unfortunately, I actually inject that one in order to use the `launchSettings.json` capabilities for IIS/IISExpress.

Alternatively, perhaps
```c#
        // CPS is Common Project System which is the new SDK style projects that DotNetCore uses
        private const string CPSCSharpCapability = "CSharp&CPS"; 
        private const string LegacyRazorEditorCapability = "LegacyRazorEditor";

        private protected virtual bool ProjectSupportsLSPEditor(string documentMoniker, IVsHierarchy hierarchy)
        {
            if (hierarchy == null)
            {
                var hr = _vsUIShellOpenDocument.Value.IsDocumentInAProject(documentMoniker, out var uiHierarchy, out _, out _, out _);
                hierarchy = uiHierarchy;
                if (!ErrorHandler.Succeeded(hr) || hierarchy == null)
                {
                    return false;
                }
            }

            if (_projectHierarchyInspector.HasCapability(documentMoniker, hierarchy, LegacyRazorEditorCapability))
            {
                // CPS project that requires the legacy editor
                return false;
            }

            if (_projectHierarchyInspector.HasCapability(documentMoniker, hierarchy, CPSCSharpCapability))
            {
                // .NET Core project that supports C#
                return true;
            }

            // Not a C# .NET Core project. This typically happens for legacy Razor scenarios
            return false;
        }
```
would allow an author to add 
```xml
        <ItemGroup>
                <ProjectCapability Include="LegacyRazorEditor" />
        </ItemGroup>
```
if needed to force the legacy editor in projects that need it.

**Reply by NTaylorMullen** [reply]
https://github.com/dotnet/razor-tooling/pull/5781
